### ISSUE TITLE
Chapter4: issue in early stopping

### ISSUE THREAD

**Issue by akjain90** [ageron/handson-ml#159]
Hi,
I am not sure about the issue, but I might be wrong.
In chapter 4 Jupyter notebook, in cell 44, there is a comment which says that "continues where it left off". I think its bit misleading because as per my understanding the sgd_reg model has restarted the training, otherwise getting best_epoch as 239 which is similar to the previous model looks infeasible (even after equating the best_epoch to None.)

Thank you

regards,
Aishvarya

**Comment by ageron** [comment]
Oops, sorry for the late response, somehow I missed this issue.

When you call `fit()` on a model, by default it forgets everything it learned before, and it just starts from scratch. However, if you set `warm_start=True` (which is possible in some estimators, such as the `SGDRegressor`), then calling `fit()` just continues training where it left off.  In this case, notice that I set `warm_start=True`:

```python
sgd_reg = SGDRegressor(max_iter=1, warm_start=True, penalty=None,
                       learning_rate="constant", eta0=0.0005, random_state=42)
```

This means that it's the same model that continues getting trained at every epoch. Its performance keeps improving (measured on the validation set) until it reaches epoch 239, and then it keeps getting trained, but it never beats the best performance again, so the `best_model` ends up being a clone of the model at epoch 239.

I hope this help.

**Comment by akjain90** [comment]
Thanks, I think I misunderstood something, things are clear now.
